frontend/components/viewer/js/viewer.service.ts

```diff
@@ -322,12 +322,18 @@ export class ViewerService {
 		}
 	}
 
-	public getScreenshot(promise) {
-		if (promise) {
-			this.initialised.promise.then(() => {
-				this.viewer.getScreenshot(promise);
-			});
-		}
+	public async getScreenshot() {
+		await this.initialised.promise;
+
+		// This is anti-pattern, but first we need to modify unity-util to use proper solution
```

Comment:
@carmenfan I can modify `unity-util` if you explain to me how can we generate new `.unityweb` file.

Reply:
@simplypixi Whatever that's generated in .unityweb can't be changed, as that's auto-generated code by unity (i.e. we don't have direct control over the code in that and it will get overwritten everytime we bake the game)

For screenshots, we need a promise (or some form of callback) because we can only take a copy of the render buffer after the next frame. As as it stands, it takes a defer object when the web request a screenshot. Then the unity game will call [this function](https://github.com/3drepo/3drepo.io/blob/master/frontend/globals/unity-util.ts#L321) with the screenshot when it has one to give back.

There's definitely better ways to do this instead of sending a defer object, but the flexibility of the refactoring is limited to the interactions within unity-utils unfortunately.

What's your plan on that?

Reply:
Ok, I understand. ;)

The current solution is dirty, but it behaves exactly like `$q.defer`, which you used before. It's the same solution implemented in a non-angular way.
I think, in that case, is better than using callbacks to achieve a similar result. So I propose to leave this as it is. Is it ok for you?

Reply:
@simplypixi what do you think about getting the unity to return a promise instead?

The one complaint we've had with people trying to integrate our stuff is that "our promise isn't the standard promise" on the unity-utils api. So I think if we return a promise like a "standard promise" instead of making the user create one it might be a better approach?

Reply:
@carmenfan It's a definitely good idea. ;) I wanted to do that, but I understood that we can not modify this file because is generated by unity.

Reply:
You can modify unity-utils (in `frontend/globals/unity-util.ts`), you just can't modify anything in `frontend/unity` folder 😃

Reply:
Ohh... 🤦‍♂️ Awesome! I'll change it tomorrow morning. :D

Reply:
Done ;)